frontend/http/frontend.go

```diff
@@ -0,0 +1,179 @@
+// in milliseconds .
+func recordResponseDuration(action string, err error, duration time.Duration) {
+	var errString string
+	if err != nil {
+		errString = err.Error()
+	}
+
+	promResponseDurationMilliseconds.
+		WithLabelValues(action, errString).
+		Observe(float64(duration.Nanoseconds()) / float64(time.Millisecond))
+}
+
+// Config represents all of the configurable options for an HTTP BitTorrent
+// Frontend.
+type Config struct {
+	Addr            string
+	ReadTimeout     time.Duration
+	WriteTimeout    time.Duration
+	RequestTimeout  time.Duration
+	AllowIPSpoofing bool
+	RealIPHeader    string
+}
+
+// Frontend holds the state of an HTTP BitTorrent Frontend.
+type Frontend struct {
+	grace *graceful.Server
+
+	logic frontend.TrackerLogic
+	Config
+}
+
+// NewFrontend allocates a new instance of a Frontend.
+func NewFrontend(logic frontend.TrackerLogic, cfg Config) *Frontend {
+	return &Frontend{
+		logic:  logic,
+		Config: cfg,
+	}
+}
+
+// Stop provides a thread-safe way to shutdown a currently running Tracker.
+func (t *Frontend) Stop() {
+	t.grace.Stop(t.grace.Timeout)
+	<-t.grace.StopChan()
+}
+
+func (t *Frontend) handler() http.Handler {
+	router := httprouter.New()
+	router.GET("/announce", t.announceRoute)
+	router.GET("/scrape", t.scrapeRoute)
+	return router
+}
+
+// ListenAndServe listens on the TCP network address t.Addr and blocks serving
+// BitTorrent requests until t.Stop() is called or an error is returned.
+func (t *Frontend) ListenAndServe() error {
+	t.grace = &graceful.Server{
+		Server: &http.Server{
+			Addr:         t.Addr,
+			Handler:      t.handler(),
+			ReadTimeout:  t.ReadTimeout,
+			WriteTimeout: t.WriteTimeout,
+		},
+		Timeout:          t.RequestTimeout,
+		NoSignalHandling: true,
+		ConnState: func(conn net.Conn, state http.ConnState) {
+			switch state {
+			case http.StateNew:
+				//stats.RecordEvent(stats.AcceptedConnection)
+
+			case http.StateClosed:
+				//stats.RecordEvent(stats.ClosedConnection)
+
+			case http.StateHijacked:
+				panic("http: connection impossibly hijacked")
+
+			// Ignore the following cases.
+			case http.StateActive, http.StateIdle:
+
+			default:
+				panic("http: connection transitioned to unknown state")
+			}
+		},
+	}
+	t.grace.SetKeepAlivesEnabled(false)
+
+	if err := t.grace.ListenAndServe(); err != nil {
+		if opErr, ok := err.(*net.OpError); !ok || (ok && opErr.Op != "accept") {
+			panic("http: failed to gracefully run HTTP server: " + err.Error())
+		}
+	}
+
+	return nil
+}
+
+// announceRoute parses and responds to an Announce by using t.TrackerLogic.
+func (t *Frontend) announceRoute(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	var err error
+	start := time.Now()
+	defer recordResponseDuration("announce", err, time.Since(start))
+
+	req, err := ParseAnnounce(r, t.RealIPHeader, t.AllowIPSpoofing)
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+
+	resp, err := t.logic.HandleAnnounce(context.TODO(), req)
```

Comment:
We're still using `context.TODO()` for these - we could simply use `context.Background` I think